### ISSUE TITLE
[BUG]: Incorrect results for fused BF16 elementwise expressions on SM120: one-graph (x * y) + z disagrees with PyTorch, split-graph version matches exactly

### ISSUE THREAD

**Issue by rolson24** [modular/modular#6159]
### Bug description

  On an SM120 GPU (`RTX 5090`), MAX produces incorrect results for a fused BF16 elementwise expression inside a single graph, while the equivalent split-across-graphs version matches
  PyTorch exactly.

  The smallest repro is:

  `(x * y) + z`

  What is surprising is:

  - the multiply alone matches PyTorch
  - the split version (`mul` in graph 1, `add` in graph 2) matches PyTorch
  - only the fused single-graph version is wrong

  This looks like a graph compiler / GPU pointwise lowering issue.

 

### Steps to reproduce

Minimal repro
```python
import numpy as np, torch
from max.driver import Accelerator, Buffer
from max.dtype import DType
from max.engine import InferenceSession
from max.graph import DeviceRef, Graph, TensorType, ops

n = 16384
dev = DeviceRef.GPU()
tt = TensorType(DType.float32, [n], device=dev)

rng = np.random.default_rng(1234)
x = rng.standard_normal(n, dtype=np.float32)
y = rng.standard_normal(n, dtype=np.float32)
z = rng.standard_normal(n, dtype=np.float32)

# fused: (x * y) + z in one graph
s1 = InferenceSession(devices=[Accelerator()])
with Graph("fused", input_types=(tt, tt, tt)) as g:
    xi, yi, zi = g.inputs
    xb = ops.cast(xi.tensor, DType.bfloat16)
    yb = ops.cast(yi.tensor, DType.bfloat16)
    zb = ops.cast(zi.tensor, DType.bfloat16)
    mul = xb * yb
    out = mul + zb
    g.output(ops.cast(mul, DType.float32), ops.cast(out, DType.float32))
m1 = s1.load(g)
mul_fused, out_fused = [o.to_numpy() for o in m1.execute(
    Buffer.from_numpy(x).to(m1.input_devices[0]),
    Buffer.from_numpy(y).to(m1.input_devices[1]),
    Buffer.from_numpy(z).to(m1.input_devices[2]),
)]

# split: graph1 does mul, graph2 does add
s2 = InferenceSession(devices=[Accelerator()])
with Graph("mul_only", input_types=(tt, tt)) as g:
    xi, yi = g.inputs
    g.output(ops.cast(
        ops.cast(xi.tensor, DType.bfloat16) *
        ops.cast(yi.tensor, DType.bfloat16),
        DType.float32,
    ))
mul_only = s2.load(g)

s3 = InferenceSession(devices=[Accelerator()])
with Graph("add_only", input_types=(tt, tt)) as g:
    mi, zi = g.inputs
    out = ops.cast(mi.tensor, DType.bfloat16) + ops.cast(zi.tensor, DType.bfloat16)
    g.output(ops.cast(out, DType.float32))
add_only = s3.load(g)

mul_split = mul_only.execute(
    Buffer.from_numpy(x).to(mul_only.input_devices[0]),
    Buffer.from_numpy(y).to(mul_only.input_devices[1]),
)[0].to_numpy()
out_split = add_only.execute(
    Buffer.from_numpy(mul_split.astype(np.float32)).to(add_only.input_devices[0]),
    Buffer.from_numpy(z).to(add_only.input_devices[1]),
)[0].to_numpy()

# torch reference
xt = torch.tensor(x, dtype=torch.bfloat16, device="cuda")
yt = torch.tensor(y, dtype=torch.bfloat16, device="cuda")
zt = torch.tensor(z, dtype=torch.bfloat16, device="cuda")
mul_ref = (xt * yt).float().cpu().numpy()
out_ref = ((xt * yt) + zt).float().cpu().numpy()

print("mul_fused diff:", np.max(np.abs(mul_fused - mul_ref)))
print("mul_split diff:", np.max(np.abs(mul_split - mul_ref)))
print("out_fused diff:", np.max(np.abs(out_fused - out_ref)))
print("out_split diff:", np.max(np.abs(out_split - out_ref)))
```
## Observed output

mul_fused diff: 0.0
mul_split diff: 0.0
out_fused diff: 0.0625
out_split diff: 0.0

## Expected behavior

The fused single-graph version and the split-graph version should agree with each other and with PyTorch BF16 semantics, or at least only differ within expected/documented tolerance.

## Actual behavior

- x * y matches PyTorch in both versions
- split mul -> add matches PyTorch exactly
- fused one-graph (x * y) + z does not

## Important counterexample

This does not appear to be explained by a simple intermediate BF16 -> FP32 -> BF16 round-trip difference.

I also tested a one-graph variant that explicitly forces:
```python
mul_recast = ops.cast(ops.cast(xb * yb, DType.float32), DType.bfloat16)
out = mul_recast + zb
```
That still matches the broken fused result, not the correct split result.

So the issue appears to be tied to the graph/codegen boundary itself, not just an extra cast round-trip.

### System information

```shell
- GPU: `NVIDIA GeForce RTX 5090`
- Architecture: `SM120`
- CUDA libs: `CUDA 13.0`
- MAX: nightly
```

**Comment by JoeLoser** [comment]
Thanks for the detailed bug report — the repro is very thorough and the counterexample with explicit casts was a helpful as well. That said, I had someone from the team investigate this and the behavior is actually working as intended, though we understand why it looks wrong. Here's what's happening:

## Root cause: FMA (fused multiply-add) contraction

When `(x * y) + z` is compiled within a single graph, the compiler contracts the multiply and add into a single FMA instruction. An FMA computes `x * y + z` with a single rounding at the end, keeping the intermediate product at full precision.

PyTorch, by contrast, rounds twice — once after the multiply (truncating back to BF16), then again after the add. Because BF16 has only 7 mantissa bits, that intermediate rounding is lossy, and the two approaches produce different results.

**Why the split-graph version matches PyTorch:** When the ops are in separate graphs/kernels, the intermediate BF16 value must be written to GPU memory between them, which forces the intermediate rounding. The compiler cannot contract ops across kernel boundaries.

**Why the explicit `cast(cast(mul, fp32), bf16)` doesn't help:** Within a single fused kernel, the compiler can see through the intermediate truncation/extension and still contract to an FMA.

**The FMA result is actually more accurate, not less** — it preserves more precision in the intermediate product. The max error you observed (`0.0625 = 2⁻⁴`) is within the BF16 ULP range for inputs of the magnitudes in your test. If you compare both results against a high-precision (FP64) reference, the FMA version will be closer.

**Not SM120-specific:** We've confirmed the same behavior on SM90 (H100) and SM100 (B200) — it's a property of the FMA contraction, not a specific GPU architecture.

## Our stance

We consider this a deliberate design choice — FMA contraction gives both better performance and better numerical accuracy. We don't aim for bitwise-exact matching with PyTorch, but rather to produce results that are at least as accurate.

That said, we recognize this is surprising and should be documented. We'll add a note about FMA contraction semantics to the MAX documentation.

Thanks again for reporting this — it's a great catch and the kind of precision analysis that helps us improve our docs. Let us know if you have any further questions.

**Comment by rolson24** [comment]
Thanks for looking into this!
That make sense. I originally thought this numerical mis-match was causing GPT-OSS-20b to not work well on SM120, but it turned out to be mostly from the flash attention implementation which is specific to sm100/90 and was not falling back to a naive version on SM120. I opened a PR for that here #6209
